### ISSUE TITLE
Comments (or comment settings?) as a field (or referenced entities?).

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2310]
This is a generic question from an almost newcommer in Backdrop/Drupal7/8: _Are comments a field, a content type setting or an entity?_. According to [this change record for D8](https://www.drupal.org/node/2100015):

> Comment settings are now a field. Comments allowed on any entity type.
> 
> **Description:** 
> 
> In Drupal 7, comment settings were per node type. The settings were managed from the 'edit node type' form and saved as variables.
> 
> In Drupal 8, comment settings are a field. This means you can add commenting to any entity type by adding a comments field. This also means you can have multiple comment threads on a single entity type so you could potentially have a 'for' and an 'against' comment field. Or a 'public' and 'admin' comments. 

Perhaps once Reference gets in Backdrop core, we might want to make them referenced entities (same as tags are term reference fields). I don't know. Just a though and topic for discussion...

Not sure if this needs to actually happen on the Field API level or if we could/should implement something like a form alter in order to just move **the comment settings** out of the vtab on the content type config page and into the manage fields/display tabs (the same way it is proposed to be done for the content type title/label settings in #1950 and #1441). Either way, it makes more sense to me UX-wise and would simplify the content type config form: one less vtab, one less set of settings for the user to configure. If they want comments enabled on their content type, they can add a comment field (or pseudo-field). The respective "Comments per page", "Show comment replies in a threaded list", "Allow comment title" etc. settings for comments could be moved in the manage display tab.

There would be challenges of course:

"Open" would be having a comment field added to the content type. What would "closed" be though? ...not having added a comment field in the first place? What if comments were open for a node and then closed? Would we have to put per-node field permissions in place (which we don't have in core)? "Hidden" would be brought back as an extra (moving the comment field to the hidden section of the manage display tab).

If we decide to actually make comments a field (as opposed to having comment **settings** be a pseudo-field), then it wouldn't make sense for them to have only one value neither would it make sense to have them limited to any number - they'd always have to be unlimited.

Anyways, D8 has made comment settings a field, in contrib land there's [Comments Entity](https://www.drupal.org/project/comments_entity) (only 15 active installations) and [Comment as field](https://www.drupal.org/sandbox/larowlan/1790736) (a sandbox with last activity 3 years ago), don't know if we need or want to do anything about it. Lets discuss...

**Comment by olafgrabienski** [comment]
> (...) move the comment settings out of the vtab on the content type config page and into the manage fields/display tabs (...) makes more sense to me UX-wise and would simplify the content type config form: one less vtab, one less set of settings for the user to configure. If they want comments enabled on their content type, they can add a comment field (or pseudo-field).

While comments as a field might be more flexible, UX-wise it makes more sense to leave them in the vertical tabs in my opinion. In difference to the title, I don't think that comments are perceived intuitively as a field from many users. And I guess it requires more steps to add and configure a comment field than to just configure comments in a vtab, don't you think so?

**Comment by klonos** [comment]
> ...don't you think so?

Yeah. Besides, the rest of the form fields are to be edited from the content create/edit form. Comments on the other hand are possible to be added while viewing content.

**Comment by klonos** [comment]
So perhaps that rules out the "comment settings as a field" proposal on the basis of better UX (less clicks to configure). What about the comments as referenced entities then (once we have Reference module on core that is)?

**Comment by mikemccaffrey** [comment]
> Also, perhaps this is confusing for users. I know that in D7 comments are entities (right?), but there is no way to edit them separately AFAIK. They are instead accessible through the "Comment fields" and "Comment display" tabs in content types. Why are we special-casing comments and you are not allowed to do the same with other entities?

I think this is the biggest problem with comments right now. They should be able to be added to any entity type, and new comment bundles should only have to be created if you actually need a different set of fields on them or such.

> Perhaps once Reference gets in Backdrop core, we might want to make them referenced entities (same as tags are term reference fields). I don't know. Just a though and topic for discussion...

This is possible, but it would require adding some additional features into the Reference module that it doesn't currently support. One big thing would be to provide a form to add new items after displaying the list of ones that are already there. However, that might be a feature we should add anyway.

Comments might actually be a good base case for what we are talking about in #100 with creating a series of pageless nodes, since they don't have a page of their own.

**Comment by klonos** [comment]
> Comments might actually be a good base case for what we are talking about in #100 with creating a series of pageless nodes, since they don't have a page of their own.

Ha, never thought of comments that way, but you are totally right.

**Comment by Al-Rozhkov** [comment]
Surprisingly, if comments will become fields (just like in D8), we have issue with page-less nodes almost solved. Some node types can be used as blocks, others can be used in regular fashion. Because nodes are almost perfect blocks. Nodes have revisions, translatable, fieldable. The only unnecessary part is comments tied to each node type.

Right now I can't even hide comments tab in vertical tabs on node add form. But it is possible in D7. Do we have issue for that functionality?

**Comment by cptX** [comment]
Please don't you ever ever ever make comments entities!!! D8/D9 did this and created thousand of new problems with comments in multilingual sites. Comments in a multilingual site actually now in D8/D9 don't work!!! I spent almost a month trying to fix as many issues as possible and still the situation is unsolvable. Actually, that's why I came to backdrop, because of the broken comment system of D9 in multilingual. At the moment in D9 the following problems exist with comments:
1. Cannot filter comments per language (so in a translated page you get mixed language comments)
2. Preview doesn't work
3. Delete doesn't work
4. Comment counter per node per language doesn't work
5. Permalinks of comments don't work if you try to patch the above issues...
